### ISSUE TITLE
LLVM error at optim level 3 (Windows and Linux 64 bit)

### ISSUE THREAD

**Issue by scharris** [rust-lang/rust#32979]
Rust 1.7 and 1.8, Windows and Linux 64 bit

At optimization level 3, build of example project (1) is exiting with an LLVM error:

cargo build --release

> Instruction does not dominate all uses!
>   <badref> = bitcast void <badref> to i8*
>   call void @llvm.lifetime.end(i64 8, i8* <badref>)
> LLVM ERROR: Broken function found, compilation aborted!

(1) https://gitlab.com/scharris/rust-compiler-error.git 

At level 2, compilation completes normally.

**Comment by scharris** [comment]
PS: I apologize for the large size of the example, I did spend quite some time cutting it down to this size.  Cutting more tends to make the problem go away.

**Comment by alexcrichton** [comment]
Looks like this is related to LTO but not related to Windows (I've reproduced on Linux). Bugs like this are typically in upstream LLVM as well, so it likely points to something there that needs fixing (but not always!).

I managed to reduce it to:

```rust
// sdar.rs
use std::str::FromStr;

pub struct TrainingIterResult {
    _a: (),
}

pub fn do_training_iters() -> Vec<TrainingIterResult> {
    Vec::new()
}

pub struct MolsFeaturesTable(());

impl MolsFeaturesTable {
    fn new(mol_ids: Vec<String>) -> MolsFeaturesTable {
        if mol_ids.len() != 3 {
            panic!()
        }

        MolsFeaturesTable(())
    }
}

pub fn read_with_activities()
    -> Result<(Vec<f32>, MolsFeaturesTable), MolsFeaturesTableError>
{
    let _ = f32::from_str("32");
    for occ_str in " 3".split(' ') {
        try!(f32::from_str(occ_str).map_err(|_| parse_err(format!(""))));
    }

    Ok((Vec::new(), MolsFeaturesTable::new(Vec::new())))
}

fn parse_err(s: String) -> MolsFeaturesTableError {
    MolsFeaturesTableError::Parse(s)
}

pub enum MolsFeaturesTableError {
    _Foo(Box<u32>),
    Parse(String),
}

// foo.rs
extern crate sdar;

fn main() {
    let _ = sdar::read_with_activities();
    sdar::do_training_iters();
}
```

Which when compiled as:

```
rustup run nightly rustc sdar.rs --crate-type lib -C opt-level=3 
rustup run nightly rustc foo.rs -L. -C opt-level=3 -Clto
```

will generate:

```
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp:531: void {anonymous}::PromoteMem2Reg::run(): Assertion `isAllocaPromotable(AI) && "Cannot promote non-promotable alloca!"' failed.
```

Note that LLVM assertions are only enabled on nightly, so you may get a different error (or not at all) on stable, but that is the same error from the original source on nightly as well.

**Comment by not-fl3** [comment]
Hello! I got same error with rust-1.9.0-stable, 1.10-nightly, 1.11-nightly, on different linux boxes(tried on gentoo and debian).

> rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp:531: void {anonymous}::PromoteMem2Reg::run(): Assertion `isAllocaPromotable(AI) && "Cannot promote non-promotable alloca!"' failed.

Way to reproduce:
- this [repo](https://github.com/not-fl3/AllocaFail). 
- run `cargo run --release` in foo folder

**Comment by alexbool** [comment]
What a nasty bug, reproduced in one of my projects too. Any info on it?

**Comment by alexbool** [comment]
In my particular case it was fixed after LLVM upgrade I think

**Comment by scharris** [comment]
This appears to have been fixed at 1.11, both for my original code and Alex's reduced example. Closing.
